kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/cloudevent/CloudEventKafkaMessageConverter.java

```diff
@@ -0,0 +1,463 @@
+        this.metadataNameResolver = builder.metadataToExtensionMap.entrySet()
+                                                                  .stream()
+                                                                  .collect(Collectors.toMap(Map.Entry::getValue,
+                                                                                            Map.Entry::getKey));
+        this.sourceSupplier = builder.sourceSupplier;
+        this.dataContentTypeSupplier = builder.dataContentTypeSupplier;
+        this.dataSchemaSupplier = builder.dataSchemaSupplier;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link CloudEventKafkaMessageConverter}.
+     * <p>
+     * The {@link SequencingPolicy} is defaulted to an {@link SequentialPerAggregatePolicy}. The {@link Serializer} is
+     * a
+     * <b>hard requirement</b> and as such should be provided.
+     *
+     * @return a Builder to be able to create a {@link CloudEventKafkaMessageConverter}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Note that the {@link ProducerRecord} created through this method sets the {@link ProducerRecord#timestamp()} to
+     * {@code null}. Doing so will ensure the used Producer sets a timestamp itself for the record. The
+     * {@link EventMessage#getTimestamp()} field is however still taken into account, but as via the CloudEvent.
+     * <p>
+     * Additional note that the ProducerRecord will be given a {@code null} {@link ProducerRecord#partition()} value. In
+     * return, the {@link ProducerRecord#key()} field is defined by using the configured {@link SequencingPolicy} to
+     * retrieve the given {@code eventMessage}'s {@code sequenceIdentifier}. The combination of a {@code null} partition
+     * and the possibly present or empty key will define which partition the Producer will choose to dispatch the record
+     * on.
+     *
+     * @see ProducerRecord
+     */
+    @Override
+    public ProducerRecord<String, CloudEvent> createKafkaMessage(EventMessage<?> eventMessage, String topic) {
+        SerializedObject<byte[]> serializedObject = eventMessage.serializePayload(serializer, byte[].class);
+        return new ProducerRecord<>(
+                topic, null, null, recordKey(eventMessage),
+                toCloudEvent(eventMessage, serializedObject),
+                null
+        );
+    }
+
+    private CloudEvent toCloudEvent(EventMessage<?> message, SerializedObject<byte[]> serializedObject) {
+        CloudEventBuilder builder = new CloudEventBuilder();
+        builder.withId(message.getIdentifier());
+        builder.withData(serializedObject.getData());
+        dataContentTypeSupplier.apply(message).ifPresent(builder::withDataContentType);
+        dataSchemaSupplier.apply(message).ifPresent(builder::withDataSchema);
+        builder.withSource(sourceSupplier.apply(message));
```

Comment:
This needs to be an URI-reference, but by default we're putting in the event type. Shouldn't we default this to something else?

Reply:
Open for suggestions, but I don't think there is a better target.

Reply:
The [spec](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md) states the following about this:

> Source
> The "source" is the context in which the occurrence happened. In a distributed system it might consist of multiple [Producers](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#producer). 
> If a source is not aware of CloudEvents, an external producer creates the CloudEvent on behalf of the source.

So based on that, I'd wager the FQCN of an Event is incorrect too. 
The FQCN without the class name attached might be an option.

More ideal would be the application name I guess?
Or, we could default to a descriptor of the running application as we do for the `JpaTokenStore`/`JdbcTokenStore`, with `ManagementFactory.getRuntimeMXBean().getName()`.

Reply:
But we need the 'original' source. The application converting the Axon Event message to a Cloud Event might not be the same as the application which created the Axon Event. We could make it mandatory, and not set a default?

Reply:
Or for example `org.axonframework.messaging` as default, so it's clear it's 'source' is from an Axon Applicaton.

Reply:
Gotcha, that's true. The app using this converter is not necessarily the origin.
 It resonates most with "context" for me, by the way. 
 However, as AF does not actively impose a context at all, that option doesn't fly.
 
 It's something we can digest within the AF5 planning, though. 
 I think we can get benefit from carrying that information on several levels, so.
 Although, that's not a solution for now.
 
 Is there _anything_ that the Cloud Events specification describes as a reasonable default in these scenarios?
 I assume we're not the only ones doing such a translation step, so.

Reply:
Based on this discussion, I'd suggest `https://www.axoniq.io/$Context/$SimpleClassName`.

I suggest using the simple class name (not the FQCN), since it might be exported to application where the java packages have no meaning at all.

Reply:
But that would be an url that doesn't resolve. Also, the class is not the source, and already there as type.